Added possibility to add listener for connectorMarkedDirty

caalador opened the pull request:
Now there is the possibility to add a listener to the UI
that is fired when ever a connector is marked as dirty.
Only the first markDirty for a connector will create an event
and no events will be fired untill it has been marked clean.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/10773)
<!-- Reviewable:end -->

caalador committed:
Added possibility to add listener for connectorMarkedDirty

Now there is the possibility to add a listener to the UI
that is fired when ever a connector is marked as dirty.
Only the first markDirty for a connector will create an event
and no events will be fired untill it has been marked clean.

Legioth commented:
*[a discussion](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L8lPLtD7C2dHIAc0MdR:-L8lPLtD7C2dHIAc0MdS:b-neo385) (no related file):*
I'm wondering whether this API is important enough to bloat the `UI` namespace, or if all the API should instead be moved to `ConnectorTracker`?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10773)*
<!-- Sent from Reviewable.io -->

caalador committed:
Move Dirty listener out from the UI to ConnectorTracker.

caalador committed:
If no listeners mapped then return empty list.

Fixes npe that shouldn't happen.

Legioth commented:
Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[server/src/main/java/com/vaadin/ui/ConnectorTracker.java, line 79 at r3](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9EO3Qa-7Xq54ZxQQvm:-L9EO3Qa-7Xq54ZxQQvn:b-ffbw5v) ([raw file](https://github.com/vaadin/framework/blob/df07fbcf8e2b68da4cf65fb1c27e1350a4df5320/server/src/main/java/com/vaadin/ui/ConnectorTracker.java#L79)):*
> ```Java
>     private final Set<ClientConnector> uninitializedConnectors = new HashSet<>();
> 
>     private HashMap<Class<?>, List<?>> listeners;
> ```

What's the purpose of this kind of complexity compared to simply having a `List<MarkedAsDirtyListener>` field?

---

*[server/src/main/java/com/vaadin/ui/ConnectorTracker.java, line 936 at r3](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9EOIMY4vCyS3lbdd3u:-L9EOIMY4vCyS3lbdd3v:b-h05k4m) ([raw file](https://github.com/vaadin/framework/blob/df07fbcf8e2b68da4cf65fb1c27e1350a4df5320/server/src/main/java/com/vaadin/ui/ConnectorTracker.java#L936)):*
> ```Java
>      */
>     public void notifyMarkedAsDirtyListeners(ClientConnector connector) {
>         getRegisteredListeners(MarkedAsDirtyListener.class).forEach(
> ```

I suspect this will fail with a `ConcurrentModificationException` if a listener removes itself. To fix this, a copy of the backing list should be iterated instead of iterating over the "live" list.

---

*[server/src/main/java/com/vaadin/ui/ConnectorTracker.java, line 938 at r3](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9EOF7q0B3UaplqyQg7:-L9EOF7q0B3UaplqyQg8:bt3lyvc) ([raw file](https://github.com/vaadin/framework/blob/df07fbcf8e2b68da4cf65fb1c27e1350a4df5320/server/src/main/java/com/vaadin/ui/ConnectorTracker.java#L938)):*
> ```Java
>         getRegisteredListeners(MarkedAsDirtyListener.class).forEach(
>                 listener -> listener.connectorMarkedAsDirty(
>                         new MarkedAsDirtyConnectorEvent(connector, uI)));
> ```

Why is a new event instance created for each listener?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10773)*
<!-- Sent from Reviewable.io -->

caalador committed:
Added missing headers.
Simplify listener handling as there probably won't
be any extra listeners added here.

caalador committed:
Remove header from test class.

Add a test that for multiple markAsDirty calls only one
handler call is fired on a roundtrip.

caalador commented:
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[server/src/test/java/com/vaadin/tests/event/MarkAsDirtyListenerTest.java, line 15 at r1](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9EiYGd8nLx2r8dBbZA:-L9EmC1z53FyycdMwWVN:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/373e69f8a3ad98274ad0f9d69a9c75c854d67e5b/server/src/test/java/com/vaadin/tests/event/MarkAsDirtyListenerTest.java#L15)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Please do not add the copyright header to test files.
</blockquote></details>

Done.

---

*[server/src/test/java/com/vaadin/tests/event/MarkAsDirtyListenerTest.java, line 38 at r1](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9Ej1C14_sCVWb8QzTa:-L9EnE9b5MIFnlO_5W5X:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/373e69f8a3ad98274ad0f9d69a9c75c854d67e5b/server/src/test/java/com/vaadin/tests/event/MarkAsDirtyListenerTest.java#L38)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Please add tests for multiple `markDirty()`calls on the same connector
The listener should be called only once, IMO
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10773)*
<!-- Sent from Reviewable.io -->

elmot commented:
Reviewed 3 of 3 files at r1, 4 of 5 files at r2, 3 of 3 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10773)*
<!-- Sent from Reviewable.io -->

tsuoanttila commented:
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[server/src/main/java/com/vaadin/event/MarkedAsDirtyConnectorEvent.java, line 23 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FCsGGD9fsmN1iGx7V:-L9FCsGGD9fsmN1iGx7W:b-7jfxkb) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/event/MarkedAsDirtyConnectorEvent.java#L23)):*
> ```Java
> /**
>  * Event which is fired for all registered MarkDirtyListeners when a
>  * connector is marked as dirty.
> ```

since

---

*[server/src/main/java/com/vaadin/event/MarkedAsDirtyConnectorEvent.java, line 42 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FChXz2Og6Kshr7s--:-L9FChXz2Og6Kshr7s-0:b-d0wxhj) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/event/MarkedAsDirtyConnectorEvent.java#L42)):*
> ```Java
>         return ui;
>     }
> }
> ```

new line

---

*[server/src/main/java/com/vaadin/event/MarkedAsDirtyListener.java, line 19 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FCrQ-3BmfRWDsIN8s:-L9FCrQ-3BmfRWDsIN8t:b-7jfxkb) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/event/MarkedAsDirtyListener.java#L19)):*
> ```Java
> 
> /**
>  * An interface used for listening to marked as dirty events.
> ```

since

---

*[server/src/main/java/com/vaadin/event/MarkedAsDirtyListener.java, line 31 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FCimt-eSjrHYELXLN:-L9FCimt-eSjrHYELXLO:b-d0wxhj) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/event/MarkedAsDirtyListener.java#L31)):*
> ```Java
>      */
>     void connectorMarkedAsDirty(MarkedAsDirtyConnectorEvent event);
> }
> ```

new line

---

*[server/src/main/java/com/vaadin/ui/ConnectorTracker.java, line 920 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FCpIh5lUEhF7e5jZ2:-L9FCpIh5lUEhF7e5jZ3:b-7jfxkb) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/ui/ConnectorTracker.java#L920)):*
> ```Java
>      * @param listener
>      *         listener to add
>      * @return registration for removing listener registration
> ```

since

---

*[server/src/main/java/com/vaadin/ui/ConnectorTracker.java, line 932 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FCq7385gAMC-MduAJ:-L9FCq7385gAMC-MduAK:b-7jfxkb) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/ui/ConnectorTracker.java#L932)):*
> ```Java
>      * has been marked as dirty.
>      *
>      * @param connector
> ```

since

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10773)*
<!-- Sent from Reviewable.io -->

caalador committed:
Add JavaDoc since.
Add missin newline to end of file.

caalador commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[server/src/main/java/com/vaadin/event/MarkedAsDirtyConnectorEvent.java, line 23 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FCsGGD9fsmN1iGx7V:-L9FHLXQC6Bq48d5_4Sq:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/event/MarkedAsDirtyConnectorEvent.java#L23)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

since
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/MarkedAsDirtyConnectorEvent.java, line 42 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FChXz2Og6Kshr7s--:-L9FHM533-74ZF__xXeU:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/event/MarkedAsDirtyConnectorEvent.java#L42)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

new line
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/MarkedAsDirtyListener.java, line 19 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FCrQ-3BmfRWDsIN8s:-L9FHVmt2Jsz-FIytzqD:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/event/MarkedAsDirtyListener.java#L19)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

since
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/event/MarkedAsDirtyListener.java, line 31 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FCimt-eSjrHYELXLN:-L9FHVXg6Wfo7nW0-O6G:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/event/MarkedAsDirtyListener.java#L31)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

new line
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/ConnectorTracker.java, line 920 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FCpIh5lUEhF7e5jZ2:-L9FHd0D7EeCRkwX19jY:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/ui/ConnectorTracker.java#L920)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

since
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/ConnectorTracker.java, line 932 at r5](https://beta.reviewable.io/reviews/vaadin/framework/10773#-L9FCq7385gAMC-MduAJ:-L9FHckbAD6S0N5AK0x7:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/2bf79a2af3f76a2704dbeea5e3851e43d359ea67/server/src/main/java/com/vaadin/ui/ConnectorTracker.java#L932)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

since
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10773)*
<!-- Sent from Reviewable.io -->

tsuoanttila commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10773)*
<!-- Sent from Reviewable.io -->
